servant-server/src/Servant/Server/Internal/Router.hs

```diff
@@ -51,43 +64,139 @@ choice (WithRequest router1) router2 =
+--
+-- Assumes that the request or text being passed
+-- in 'WithRequest' or 'DynamicRouter' does not
+-- affect the structure of the underlying tree.
+--
+routerStructure :: Router' a -> RouterStructure
+routerStructure (WithRequest f) =
+  WithRequestStructure $
+    routerStructure (f (error "routerStructure: dummy request"))
+routerStructure (StaticRouter m ls) =
+  StaticRouterStructure (fmap routerStructure m) (length ls)
+routerStructure (DynamicRouter f) =
+  DynamicRouterStructure $
+    routerStructure (f (error "routerStructure: dummy text"))
+routerStructure (RawRouter _) =
+  RawRouterStructure
+routerStructure (Choice r1 r2) =
+  ChoiceStructure
+    (routerStructure r1)
+    (routerStructure r2)
+
+-- | Compare the structure of two routers.
+--
+sameStructure :: Router' a -> Router' b -> Bool
+sameStructure r1 r2 =
+  routerStructure r1 == routerStructure r2
+
+-- | Provide a textual representation of the
+-- structure of a router.
+--
+routerLayout :: Router' a -> Text
+routerLayout router =
+  T.unlines (["<root>"] ++ mkRouterLayout False (routerStructure router))
+  where
+    mkRouterLayout :: Bool -> RouterStructure -> [Text]
+    mkRouterLayout c (WithRequestStructure r)    = mkRouterLayout c r
+    mkRouterLayout c (StaticRouterStructure m n) = mkSubTrees c (M.toList m) n
+    mkRouterLayout c (DynamicRouterStructure r)  = mkSubTree c "<dyn>" (mkRouterLayout False r)
+    mkRouterLayout c  RawRouterStructure         =
+      if c then ["├─ <raw>"] else ["└─ <raw>"]
+    mkRouterLayout c (ChoiceStructure r1 r2)     =
+      ["┆"] ++ mkRouterLayout True r1 ++ ["┆"] ++ mkRouterLayout c r2
+
+    mkSubTrees :: Bool -> [(Text, RouterStructure)] -> Int -> [Text]
+    mkSubTrees _ []             0 = []
+    mkSubTrees c []             n =
+      replicate (n - 1) (mkLeaf True) ++ [mkLeaf c]
+    mkSubTrees c [(t, r)]       0 =
+      mkSubTree c    t (mkRouterLayout False r)
+    mkSubTrees c ((t, r) : trs) n =
+      mkSubTree True t (mkRouterLayout False r) ++ mkSubTrees c trs n
+
+    mkLeaf :: Bool -> Text
+    mkLeaf True  = "├─•"
+    mkLeaf False = "└─•"
+
+    mkSubTree :: Bool -> Text -> [Text] -> [Text]
+    mkSubTree True  root children = ("├─ " <> root) : map ("│  " <>) children
+    mkSubTree False root children = ("└─ " <> root) : map ("   " <>) children
+
+-- | Apply a transformation to the response of a `Router`.
+tweakResponse :: (RouteResult Response -> RouteResult Response) -> Router -> Router
+tweakResponse f = fmap (\a -> \req cont -> a req (cont . f))
+
 -- | Interpret a router as an application.
 runRouter :: Router -> RoutingApplication
 runRouter (WithRequest router) request respond =
   runRouter (router request) request respond
-runRouter (StaticRouter table) request respond =
+runRouter (StaticRouter table ls) request respond =
   case pathInfo request of
-    first : rest
-      | Just router <- M.lookup first table
+    []   -> runChoice ls request respond
+    [""] -> runChoice ls request respond
```

Comment:
This is for urls with trailing slashes, right? (Just trying to understand the code.)